### PR TITLE
Update Dropping Center Outcome Details via Daily Cron Job

### PR DESCRIPTION
**Pull request by belwalshubham**
## Description

A daily cron job executes to update the outcome Fields.

Updating Record :
1. Cash Contributions: Updates the cash contribution totals.
2. Product Sale Amounts: Updates the total product sale amounts.
3. Footfall Counts: Updates the count of visitors based on material contribution activities.
4. Vehicle Dispatch Counts: Updates the count of vehicles that dispatched materials.
5. Bags Received: Updates the total number of bags received from the center.

image

## Summary by CodeRabbit

- **New Features**
	- Introduced a new function to aggregate and update metrics related to dropping centers in CiviCRM, enhancing data tracking and reporting capabilities.
	- Added methods for sending email notifications to the Material Management Team and processing dispatch events, improving communication and operational efficiency.
	- Defined a new constant for relationship type in the Material Management Team.
- **Bug Fixes**
	- Enhanced error handling in email notification processes for better traceability.
- **Refactor**
	- Improved conditional logic and code formatting for clarity and adherence to coding standards.

### CODE REVIEW

File: wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    if (!isset($productSaleAmountById[$id])) {
+      $productSaleAmountById[$id] = 0;
+    }
+
+    // Sum the Cash Contribution.
+    $cashContributionById[$id] += (float) ($center['Donation.Cash_Contribution'] ?? 0);
+
+    // Sum the Product Sale Amount.
+    $productSaleAmountById[$id] += (float) ($center['Donation.Product_Sale_Amount_GBG_'] ?? 0);
+  }
```
[the remark this answers is not in the document]
Comment (belwalshubham) [reply wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php]: done

File: wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+  foreach ($vehicleDispatches as $dispatch) {
+    $dispatchId = $dispatch['Camp_Vehicle_Dispatch.Collection_Camp'];
+
+    if (!isset($vehicleDispatchCount[$dispatchId])) {
+      $vehicleDispatchCount[$dispatchId] = 0;
+    }
+
+    $vehicleDispatchCount[$dispatchId] += 1;
+  }
```
[the remark this answers is not in the document]
Comment (belwalshubham) [reply wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php]: done

File: wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+
+  foreach ($activities as $activity) {
+    $droppingCenterId = $activity['Material_Contribution.Dropping_Center'];
+
+    if (!isset($totalFootfall[$droppingCenterId])) {
+      $totalFootfall[$droppingCenterId] = 0;
+    }
+
+    $totalFootfall[$droppingCenterId] += 1;
+  }
```
[the remark this answers is not in the document]
Comment (belwalshubham) [reply wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php]: done

File: wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+  foreach ($bagData as $record) {
+    $dispatchId = $record['Camp_Vehicle_Dispatch.Collection_Camp'];
+    $bagsReceived = (int) ($record['Acknowledgement_For_Logistics.No_of_bags_received_at_PU_Office'] ?? 0);
+
+    if (!isset($bagsReceivedCount[$dispatchId])) {
+      $bagsReceivedCount[$dispatchId] = 0;
+    }
+    $bagsReceivedCount[$dispatchId] += $bagsReceived;
+  }
```
[the remark this answers is not in the document]
Comment (belwalshubham) [reply wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php]: done

File: wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+      ->execute();
+  }
+
+  // Update Bags Received.
+  foreach ($bagsReceivedCount as $id => $bagsReceived) {
+    EckEntity::update('Collection_Camp', TRUE)
+      ->addValue('Dropping_Center_Outcome.Total_no_of_bags_received_from_center', max($bagsReceived, 0))
+      ->addWhere('id', '=', $id)
+      ->execute();
+  }
```
[the remark this answers is not in the document]
Comment (belwalshubham) [reply wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php]: Done, thanks for this suggestion.

File: wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    $cashContributionById[$id] = ($cashContributionById[$id] ?? 0) + (float) ($center['Donation.Cash_Contribution'] ?? 0);
+    $productSaleAmountById[$id] = ($productSaleAmountById[$id] ?? 0) + (float) ($center['Donation.Product_Sale_Amount_GBG_'] ?? 0);
+  }
+
+  $vehicleDispatches = EckEntity::get('Collection_Source_Vehicle_Dispatch', TRUE)
+    ->addSelect('Camp_Vehicle_Dispatch.Collection_Camp')
+    ->addWhere('Camp_Vehicle_Dispatch.Collection_Camp', 'IS NOT NULL')
+    ->execute();
+
+  $vehicleDispatchArray = $vehicleDispatches->getIterator()->getArrayCopy();
```
Comment (tarunnjoshi) [inline wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php:35]: What we are trying to do here.

**Review by tarunnjoshi — CHANGES_REQUESTED** [review]

**Review by tarunnjoshi — APPROVED** [review]

File: wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php (hunk 6 above)
Comment (belwalshubham) [reply wp-content/civi-extensions/goonjcustom/api/v3/Goonjcustom/DroppingCenterOutcomeCron.php:35]: Converting that data into an array, and counting how many dispatch records exist for vehicle dispatch

**Comment by belwalshubham** [comment]
@pokhiii I’m merging this PR for now since Tarun has approved it. If you feel any changes are needed, we can address them in the next PR.

**Pull request merged**